src/foam/nanos/http/FileWebAgent.java

```diff
@@ -0,0 +1,125 @@
+package foam.nanos.http;
+
+import foam.core.X;
+import foam.nanos.boot.NSpec;
+import foam.nanos.boot.NSpecAware;
+import foam.util.SafetyUtil;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.text.StringEscapeUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.*;
+import java.util.HashMap;
+
+public class FileWebAgent
+    implements WebAgent, NSpecAware
+{
+  protected static final int                     BUFFER_SIZE = 4096;
+  protected static final String                  DEFAULT_EXT = "application/octet-stream";
+  protected static final HashMap<String, String> EXTS        = new HashMap<>();
+
+  static {
+    EXTS.put("js",    "application/javascript");
+    EXTS.put("json",  "application/json");
+    EXTS.put("class", "application/java-vm");
+    EXTS.put("xml",   "application/xml");
+
+    EXTS.put("gif",   "image/gif");
+    EXTS.put("png",   "image/png");
+    EXTS.put("svg",   "image/svg+xml");
+
+    EXTS.put("java",  "text/x-java-source");
+    EXTS.put("csv",   "text/csv");
+    EXTS.put("txt",   "text/plain");
+    EXTS.put("html",  "text/html");
+  }
+
+  protected NSpec nspec_;
+  protected String path_;
+  protected String cwd_;
+
+  public FileWebAgent() {
+    this("");
+  }
+
+  public FileWebAgent(String path) {
+    this(path, System.getProperty("user.dir"));
+  }
+
+  public FileWebAgent(String path, String cwd) {
+    this.path_ = path;
+    this.cwd_ = cwd;
+  }
+
+  @Override
+  public void execute(X x) {
+    PrintWriter pw = (PrintWriter) x.get(PrintWriter.class);
+    HttpServletRequest req = (HttpServletRequest) x.get(HttpServletRequest.class);
+    HttpServletResponse resp = (HttpServletResponse) x.get(HttpServletResponse.class);
+
+    String path = null;
+    BufferedInputStream is = null;
+
+    try {
+      path = req.getRequestURI().replaceFirst("/?service/" + nspec_.getName() + "/?", "") + path_;
+      File src = new File(SafetyUtil.isEmpty(path) ? "./" : path);
+      boolean pathStartsWithCwd = src.getAbsolutePath().startsWith(cwd_);
+
+      // handle reading of directories
+      if ( src.isDirectory() && src.canRead() && pathStartsWithCwd ) {
```

Comment:
Not a critical change, but I think it's a safer style to early exit if pathStartsWithCd is false, rather than including it in every serving condition.

Otherwise if someone adds a case later that forgets to include && pathStartsWithCwd then we could have a security problem.